src/lib.rs

```diff
@@ -479,39 +487,103 @@ impl CircBuf {
         let len = self.len();
         self.get_bytes_upto_size(len)
     }
+
+    /// Return a reader that doesn't advance the read cursor.
+    pub fn reader_peek(&self) -> CircBufPeekReader {
+        CircBufPeekReader {
+            inner: self,
+            peek_cursor: self.read_cursor,
+        }
+    }
+
+    unsafe fn read_peek(
+        src: &[u8],
+        dest: &mut [u8],
+        len: usize,
+        read_cursor: usize,
+        write_cursor: usize,
+    ) -> io::Result<usize> {
+        // either we have enough data to copy or we don't
+        let num_to_read = if len < dest.len() { len } else { dest.len() };
+        let num_to_end = src.len() - read_cursor;
+
+        // check if we need to wrap around the buffer to read num_to_read bytes
+        if num_to_read > num_to_end {
+            copy_nonoverlapping(&src[read_cursor], &mut dest[0], num_to_end);
+            copy_nonoverlapping(&src[0], &mut dest[num_to_end], num_to_read - num_to_end);
+        } else {
+            copy_nonoverlapping(&src[read_cursor], &mut dest[0], num_to_read);
+        }
+
+        Ok(num_to_read)
+    }
 }
 
-impl io::Read for CircBuf {
+pub struct CircBufPeekReader<'a> {
+    inner: &'a CircBuf,
+    peek_cursor: usize,
+}
+
+impl<'a> CircBufPeekReader<'a> {
+    /// Return the remaining number of bytes to peek
+    pub fn len(&self) -> usize {
+        CircBuf::cacl_len(
+            self.peek_cursor,
+            self.inner.write_cursor,
+            self.inner.buf.len(),
+        )
+    }
+
+    /// Reset the peek cursor to the original read cursor of the inner `CircBuf`
+    pub fn reset(&mut self) {
+        self.peek_cursor = self.inner.read_cursor;
+    }
+
+    /// Return the number of bytes currently peeked. Useful to use for `CircBuf::advance_read`.
+    pub fn readed(&self) -> usize {
```

Comment:
nit: What do you think about naming this just `read`?

Reply:
I don't like much `read`, it's not clear enough. `count()` ? `count_read()` (`read_count()`)? `count_peek()` ?

Reply:
Good point, maybe `count_peek` then?